File: widgets.py
#!/usr/bin/env python
import os
import time
# from action import Workspace_x

def get_current_workspace():
    worspace_num = int(os.popen("xdotool get_num_desktops").read())
    workspace = " "
    for i in range(1, worspace_num + 1):
        mouse_location = os.popen("xdotool getmouselocation").read()
        x = int(mouse_location[mouse_location.find('x')+2: mouse_location.find('y') - 1])
        y = int(mouse_location[mouse_location.find('y')+2: mouse_location.find('s') - 1])
        if x <= worspace_num * 22  and y <= 29:
            workspace_selected = (x - 20) % 22
            os.system("xdotool set_desktop " + str(workspace_selected))
            print(workspace_selected)
        current_workspace = int(os.popen("xdotool get_desktop").read())
        if i == current_workspace + 1:
            workspace += "%{F#cc99ff}%{+u}" + \
                str(current_workspace + 1) + "%{-u}%{F} "
            continue
        workspace += str(i) + " "
    return "%{c}" + workspace + "%{c}"


def get_widow_name():
    window_name = os.popen(
        "xprop -id $(xprop -root _NET_ACTIVE_WINDOW | cut -d ' ' -f 5) WM_NAME").read()
    # window_name = os.popen("xprop -id $(xprop -root _NET_ACTIVE_WINDOW | cut -d ' ' -f 5) WM_CLASS").read()
    window_name = window_name[window_name.find("\"") + 1:-2]
    if len(window_name) > 25:
        window_name = window_name[:26]
    return "%{c}" + window_name + "%{c}"


def get_music_status():
    music = os.popen(
        'playerctl metadata --format "playing: {{ artist }} - {{ album }} - {{ title }}"').read().rstrip("\n")
    if len(music) > 25:
        music = music[:26]
    return "%{c}" + music + "%{c}"


def get_net_rate():
    """ """
    down_before = float(
        os.popen("cat /proc/net/dev | grep wlp3s0 | awk   '{printf $2}'").read())
    up_before = float(
        os.popen("cat /proc/net/dev | rg wlp3s0 | awk '{print $10}'").read())
    time.sleep(1)
    down_after = float(
        os.popen("cat /proc/net/dev | grep wlp3s0 | awk   '{printf $2}'").read())
    up_after = float(
        os.popen("cat /proc/net/dev | rg wlp3s0 | awk '{print $10}'").read())
    down = (down_after - down_before) / 1024 / 1024
    up = (up_after - up_before) / 1024 / 1024
    return "%{c}" + "\uf0ed" + str(":{0}MB/s ".format("%.2f" % down)) + "\uf0ee" + str(":{0}MB/s".format("%.2f" % up)) + "%{c}"


def get_volume():
    """ """
    volume_get = str(os.popen("amixer get Master | tail -n1 ").read())
    first = volume_get.find('[')
    end = volume_get.find(']')
    volume = int(volume_get[first+1:end-1])
    if volume <= 34:
        return " \uf027" + str(volume)
    return "\uf028" + str(volume)


def clock():
    year_month_day = time.strftime("%Y-%m-%d", time.localtime())
    hour_minute = time.strftime("%H:%M", time.localtime())
    return " \uf073" + year_month_day + " " + hour_minute + " "


def get_memory():
    """ """
    mem_free = round(int(os.popen(
        "grep  'MemAvailable:' /proc/meminfo | awk '{print $2}'").read()) / 1024 / 1024, 1)
    mem_total = round(int(os.popen(
        "grep  'MemTotal:' /proc/meminfo | awk '{print $2}'").read()) / 1024 / 1024, 1)
    mem_used = round(mem_total - mem_free, 1)
    return " %{c}\uf4bc" + str(mem_used) + 'GB/' + str(mem_total) + 'GB%{c}'


def get_cpu():
    """ """
    cpu_usage = str(
        os.popen("top -bn1 | grep 'Cpu(s)' | awk '{print $2}'").read()).rstrip('\n')
    return " %{c}\uf2db" + cpu_usage + "%%%{c}"


def get_light():
    """ """
    light = str(int(float(os.popen("light").read().rstrip('\n'))))
    return " \uf0eb" + light

File: widgets.lua
#!/usr/bin/env lua
local col = require("color.color")
local widgets = {}
local lemonbar_root = "/home/hpxx/lemonbar_lua/"
widgets.workspace = {
	color = {
		foreground = col.light_blue,
		background = col.dark_blue,
	},
	action = nil,
	cmd = function()
		local workspace = " "
		local nums = io.popen("xdotool get_num_desktops"):read("n")
		for i = 1, nums do
			local current = io.popen("xdotool get_desktop"):read("n") + 1
			if i == current then
				workspace = workspace .. "%{F#" .. col.pink .. "}%{+u}" .. tostring(current) .. "%{-u}%{F} "
			else
				workspace = workspace .. tostring(i) .. " "
			end
		end
		return workspace
	end,
	stdout = lemonbar_root .. "workspace.lua",
	interval = 0.5,
	gap = 15,
	length = 220,
}
widgets.date = {
	color = widgets.workspace.color,
	action = nil,
	cmd = function()
		local date = string.gsub(io.popen("LANG=en_US date +%a-%b-%d"):read("a"), "\n", "")
		return "%{c}" .. date .. "%{c}"
	end,
	stdout = lemonbar_root .. "date.lua",
	interval = 0.5,
	gap = 5,
	length = 150,
}
widgets.time = {
	color = widgets.workspace.color,
	action = nil,
	cmd = function()
		local time = string.gsub(io.popen("LANG=en_US date +%H:%M"):read("a"), "\n", "")
		return "%{c}" .. time .. "%{c}"
	end,
	stdout = lemonbar_root .. "time.lua",
	interval = 2,
	gap = 5,
	length = 120,
}
widgets.volume = {
	color = widgets.workspace.color,
	action = {
		[4] = "amixer sset Master 5%+",
		[5] = "amixer sset Master 5%-",
	},
	cmd = function()
		local volume = io.popen("amixer get Master | tail -n1"):read("a")
		local before, _ = string.find(volume, "[", 1, true)
		local after, _ = string.find(volume, "]", 1, true)
		volume = string.sub(volume, before + 1, after - 2)
		return "" .. volume .. ""
	end,
	stdout = lemonbar_root .. "volume.lua",
	interval = 0.5,
	gap = 5,
	length = 60,
}
widgets.light = {
	color = widgets.workspace.color,
	action = {
		[4] = "light -A 5",
		[5] = "light -U 5",
	},
	cmd = function()
        local light = io.popen("light"):read("n")
		return "" .. light .. ""
	end,
	stdout = lemonbar_root .. "light.lua",
	interval = 0.5,
	gap = 5,
	length = 60,
}
return widgets
